README.md
# The LDAR Simulator
## About LDAR-Sim
The Leak Detection and Repair Simulator (LDAR-Sim) is an open-source modeling framework for exploring the effectiveness of methane leak detection programs. The purpose of LDAR-Sim is to enable transparent, collaborative, flexible, and intuitive investigation of emerging LDAR technologies, methods, work practices, regulations, and deployment strategies.

LDAR-Sim has many potential uses, including: 
  1) Test emissions reduction equivalence among distinct LDAR programs
  2) Evaluate performance and cost of methane sensing technologies and work practices
  3) Predict the emissions mitigation of proposed or existing fugitive methane policies
  4) Inform the development and niche of technologies and work practices

To learn more about LDAR-Sim, you can:
  1) Read our [story map](https://arcg.is/1rXeX10) (less technical introduction).
  2) Read [Fox et al., 2020](https://www.sciencedirect.com/science/article/pii/S0959652620352811).

For first time users, we recommend attempting to reproduce the case study results in Fox et al. 2020 (see below).

Thomas Fox: thomas@highwoodemissions.com

Mozhou Gao: mozhou.gao@ucalgary.ca

Thomas Barchyn: tbarchyn@ucalgary.ca

Chris Hugenholtz: chhugenh@ucalgary.ca

## LDAR-Sim Licensing and Use
LDAR-Sim was invented by Thomas Fox, Mozhou Gao, Thomas Barchyn, and Chris Hugenholtz at the University of Calgary's Centre for Smart Emissions Sensing Technologies. 

LDAR-Sim is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, version 3. LDAR-Sim is distributed in the hope that it will be useful, but without any warranty; without even the implied warranty of merchantability or fitness for a particular purpose. See the GNU Affero General Public License for more details.

LDAR-Sim code may not be modified without sharing the changes, pursuant to the GNU Affero GPL v3 License.

## [Fox_etal_2020 Release](https://github.com/tarcadius/LDAR_Sim/tree/Fox_etal_2020)
You are currently in the Fox et al. 2020 release branch. This is the most recent stable release and is recommended for use. To visit the master branch, which is evolving dynamically and is not a stable release, please [click here](https://github.com/tarcadius/LDAR_Sim).

This release contains the exact code and inputs used in our [LDAR-Sim synthesis paper](https://www.sciencedirect.com/science/article/pii/S0959652620352811). We recommend using this release, especially for first time users.

Citation for this release: Fox, Thomas A., Mozhou Gao, Thomas E. Barchyn, Yorwearth L. Jamin, and Chris H. Hugenholtz. "An agent-based model for estimating emissions reduction equivalence among leak detection and repair programs." Journal of Cleaner Production (2020): 125237.

### Getting started
This guide is intended to help interested parties reproduce the results from Fox et al 2020, which introduces LDAR-Sim and presents a case study demonstrating various hypothetical LDAR programs in Alberta, Canada.

#### Step 1: Before you begin
Read and understand the LDAR-Sim LICENSE (GNU Affero General Public License Version 3).
Read [Fox et al 2020](https://www.sciencedirect.com/science/article/pii/S0959652620352811) to familiarize yourself with LDAR-Sim fundamentals.

#### Step 2: Libraries and data
Install python 3.x. and ensure all required python modules/packages/libraries are available, as listed in the Pipfile.
The easiest way to prepare your python installation is to use [pipenv](https://pipenv.pypa.io/en/latest/) to manage a virtual environment that has the required packages. Navigate to the model_code directory and type: 

`pipenv install`

To make things easier, we have included windows binaries for the specific versions of cftime, GDAL, netCDF4, and pyproj. In the Pipfile, basemap is downloaded directly from [here](https://download.lfd.uci.edu/pythonlibs/s2jqpv5t/basemap-1.2.1-cp37-cp37m-win_amd64.whl). If this link breaks or the package is no longer available, this is not a problem for base operations of LDAR-Sim as basemap is only used in the live plotter demonstration.

#### Step 3: Reproduce OGI simulations
The ldar_sim_main file is currently configured for the OGI comparison case study presented in Fox et al 2020. Four different OGI-based LDAR programs are parameterized, which differ only according to weather and labour constraints.

Open and run ldar_sim_main.py - you may need to set your workding directory on line 32, but if you download the entire branch from Github, you should not need to. The working directory used in LDAR-Sim should contain all inputs (this is the case_study folder provided to you). If all input files and program files are in your working directory, no other changes should be required to run the OGI scenarios from the case study. 

The only difference between these simulations and those in the Fox et al. 2020 study is that only 3 repeat simulations are run for each program in this demonstration, whereas in the paper, 25 simulations are run for each program to constrain uncertainty. Running 3 sets of simulations for each program over multiple years, rather than 25, will take much less time. Results should resemble Figures 2C and 2D in Fox et al 2020 but will not be exactly the same as the model is stochastic.

LDAR-Sim will automatically output a set of figures and spreadsheets comparing among programs and a folder for each program. The program-specific folders will each contain exhaustive data on leaks, facilities, inspection crews, and so on for each simulation that is run.

#### Step 4: Reproduce alt-LDAR simulations
To run the alternative programs from the case study, change the program list on line 34 of ldar_sim_main to read ['P_ref', 'P_1', 'P_2']. This will change to the appropriate input files, which are already in your working directory.

The main differences among programs are presented in Table 1 of Fox et al 2020. Both aircraft and MGL (truck) programs require a second method called OGI follow-up (OGI_FU) to diagnose individual leaks and perform repairs. The reference program will be the same as Step 3 above.

#### Step 5: Explore
Once familiar with the case study, users can explore different input configurations. Just as each simulation can compare multiple LDAR programs, so can each LDAR program consist of multiple different LDAR methods. We provide generic OGI, aircraft, and truck methods, but others can be devised using generic methods as templates. Users should explore different facility inputs, methods combinations and parameterizations, empirical emissions data, and so on.

## Other versions
Several LDAR-Sim advances are not publicly available at this time, including more advanced equivalence scenario modeling, specific method modules, and cost-effectiveness comparisons.

## Contributions and collaboration
The authors welcome all contributions and collaborations. Please reach out - we would love to hear from you and/or work with you!


changelog.md
Change Log Version 2.0 (21/09):
1.	**External leak generator -** Optionally creates a single set of leaks that can be used across multiple simulations to enable "apples to apples" comparisons of different programs. Leak lists can also be saved and shared to improve reproducibility.
2.	**Modularized methods -** Methods (e.g., aircraft) are now comprised of a deployment style (platform), a sensor type, the scale of measurement (equipment vs component vs site), and whether or not they perform follow-up inspections. Old method modules are now defunct.
3.	**User manual -** Comprehensive documentation for all input parameters, including descriptions and warnings. Also includes a developer guide for those who want to contribute to the public code base.
4.	**Added satellite -** Satellite module now available, including satellite sensors, deployment strategies, and scheduling.
5.	**Crew scheduling -** Allows route-planning, specifying survey dates and date ranges, and includes route optimization based on site, home base, and airport locations. 
6.	**Company scheduling -** Companies can deploy crews strategically to spatial clusters of sites. Useful when assets are distributed across large geographical areas.
7.	**Switched to MIT license** MIT licensing will be used from version 2 onwards.
8.  **Start and end dates -** Instead of start year and a number of timesteps, specific start and end dates must be provided.
9.	**Hourly weather -** Optional ability now exists to query weather on an hourly basis rather than daily basis.
10.	**New sensitivity analysis -** Added a sensitivity analysis program called ldar_sim_sens.py that runs LDAR-Sim from a wrapper that will generate either a series of programs based on a single range of variables or a series of simulation sets based on multiple ranges of variables.
11. **Operator agent -** A new operator agent that can be assigned to a site and can use any sensor (default is AVO - auditory, visual, olfactory)
12.	**Instant follow-up -** Allows company to flag sites immediately and prioritize them for follow up if the emission rate from the site is greater than a user-defined threshold.
13.	**Follow-up delay -** Allows a company to wait a set amount of time (e.g., accumulate multiple screening measurements, even of the same site) before triaging sites for follow-up. Site redundancy can be handled using a new redundancy filter where maximum, average, or recent emission rates can be used to rank sites by severity.
14.	**Input mapper -** Maps user-defined inputs to relevant internal defaults and checks input types for issues before running simulations.
15.	**Local weather -** Added a function to convert weather data to local time zone when large regions are modeled.
16.	**Accessible weather -** Improved custom ERA5 downloader and in-file support documentation.
17. **Internal defaults -** Internal default parameters for all methods to mitigate mistakes due to previously complex parameter tracking.
18. **Rollover -** All methods now use a rollover, meaning that if a site cannot be completed before the end of the day, crews return to a home base and come back the next day to finish that site. If a site requires multiple days to inspect, crews will continue to return to that site until it is completed. This update more accurately reflects real field operations.
19.	**Labor calibration -** Option to automatically estimate the number of required crews for a given program and set minimum intervals to ensure that surveys are approximately evening distributed across each compliance period. This is performed before simulations are run and is used to estimate when a survey should be performed based on the required survey frequency and how many of crews are required to perform all surveys within the set time interval.
20. **Improved economics -** New economics functionality with cost_mitigation.py that calculates the value of natural gas mitigated with LDAR programs below baseline emissions. This value is used, along with total costs, to derive a cost mitigation ratio for each LDAR program input into the model. Automatically outputs new figures.
21. **Removed leak count and spin up -** These two inputs were redundant and prone to error. Initial emissions are now estimated using leak generation and removal assumptions.
22. **Seed timeseries generator –** Useful for testing as each day will have the same seed for all programs in a simulation set. 
23.	**Code cleanup –** Removed unused or obsolete code, including defunct sensitivity analyses.
24. **Changed to YAML inputs -**  Switched from text file inputs to YAML to improve readability.
25.	**Restructured software folder -** Improved folder structure for entire software to be more modular and follow current practices.
26.	**Updated input data structure -** An object-orientated input data structure is now used.
27. **Package upgrades -** Updated Python packages, requirements file, and pipenv files to enable flexibility in package management.
28. **Improved parameter organization -** Input parameter lists are now much longer but are better organized. New rules are also implemented to ensure that consistent data types are used for lists and that nested dictionaries are appropriately managed.


LDAR_Sim/model_code/sensitivity.py
# ------------------------------------------------------------------------------
# Program:     The LDAR Simulator (LDAR-Sim) 
# File:        Sensitivity analysis
# Purpose:     Module to enable sensitivity analysis
#
# Copyright (C) 2018-2020  Thomas Fox, Mozhou Gao, Thomas Barchyn, Chris Hugenholtz
#    
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, version 3.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# ------------------------------------------------------------------------------
import numpy as np
import pandas as pd
import os
import time


class Sensitivity:
    def __init__(self, parameters, timeseries, state):
        """
        Initialize a sensitivity analysis for a given program.

        """
        self.parameters = parameters
        self.timeseries = timeseries
        self.state = state

        # Make folder for sensitivity analysis outputs
        self.output_directory = os.path.join(self.parameters['working_directory'], 'sensitivity_analysis')
        if not os.path.exists(self.output_directory):
            os.makedirs(self.output_directory)

        # ------------------------Define SA input parameters----------------------------
        output_file = os.path.join(self.output_directory, 'SA_params.csv')
        param_df = pd.DataFrame()
        if self.parameters['simulation'] == '0' and self.parameters['sensitivity']['order'] == '1':
            for i in range(self.parameters['n_simulations']):
                row = {
                    # General inputs
                    'LSD_outliers': int(np.round(np.random.normal(0, 1))),
                    'LSD_samples': int(np.random.normal(len(self.state['empirical_leaks']), len(self.state['empirical_leaks']) / 4)),
                    'LCD_outliers': int(np.round(np.random.normal(0, 1))),
                    'LCD_samples': int(np.random.normal(len(self.state['empirical_counts']), len(self.state['empirical_counts']) / 4)),
                    'site_rate_outliers': int(np.round(np.random.normal(0, 1))),
                    'site_rate_samples': int(np.random.normal(len(self.state['empirical_sites']), len(self.state['empirical_sites']) / 4)),
                    'offsite_times_outliers': int(np.round(np.random.normal(0, 1))),
                    'offsite_times_samples': int(np.random.normal(len(self.state['offsite_times']), len(self.state['offsite_times']) / 4)),

                    'LPR': np.random.gamma(0.8327945, 0.03138632365),
                    'repair_delay': np.random.uniform(0, 100),
                    'operator_strength': np.random.exponential(0.1),
                    'max_det_op': np.random.exponential(1 / 5),
                    'consider_operator': bool(np.random.binomial(1, 0.2)),
                    'consider_daylight': bool(np.random.binomial(1, 0.5)),
                    'consider_venting': bool(np.random.binomial(1, 0.5)),
                    'max_workday': round(np.random.uniform(6, 14)),
                    'start_year': round(np.random.uniform(2003, 2012)),

                    # OGI inputs - only used if called
                    'OGI_n_crews': np.random.poisson(0.5) + 1,
                    'OGI_min_temp': np.random.normal(-20, 10),
                    'OGI_max_wind': np.random.normal(15, 3),
                    'OGI_max_precip': np.random.uniform(0, 0.1),  # in meters
                    'OGI_reporting_delay': np.random.uniform(0, 30),
                    'OGI_time': np.random.uniform(30, 500),
                    'OGI_required_surveys': np.random.uniform(1, 4),
                    'OGI_min_interval': np.random.uniform(0, 90),
                    'OGI_MDL': np.random.uniform(0, 2),

                    # MGL inputs - only used if called
                    'truck_n_crews': np.random.poisson(0.5) + 1,
                    'truck_min_temp': np.random.normal(-20, 10),
                    'truck_max_wind': np.random.normal(15, 3),
                    'truck_max_precip': np.random.uniform(0, 0.1),
                    'truck_reporting_delay': np.random.uniform(0, 30),
                    'truck_time': np.random.uniform(1, 30),
                    'truck_required_surveys': np.random.uniform(1, 4),
                    'truck_min_interval': np.random.uniform(0, 90),
                    'truck_MDL': np.random.uniform(1, 100),  # grams/hour
                    'truck_follow_up_thresh': np.random.uniform(0, 500),
                    'truck_follow_up_ratio': np.random.uniform(0.1, 1)
                }
                param_df = param_df.append(pd.DataFrame([row]))
            param_df.to_csv(output_file, index=False)

        while not os.path.exists(output_file):
            time.sleep(10)
        if os.path.isfile(output_file):
            params = pd.read_csv(output_file)
            current_row = params.iloc[int(self.parameters['simulation'])]
            self.SA_params = current_row.to_dict()


        # Go find the secret message (i.e. load required SA parameters)!
        # for attempts in range(int(1e10)):
        #     try:
        #         params = pd.read_csv(output_file)
        #         current_row = params.iloc[int(self.parameters['simulation'])]
        #         self.SA_params = current_row.to_dict()
        #         break
        #
        #     except: continue

        # --------------Update model parameters according to SA definition--------------

        self.parameters['LPR'] = self.SA_params['LPR']
        self.parameters['start_year'] = self.SA_params['start_year']
        self.parameters['repair_delay'] = self.SA_params['repair_delay']
        self.parameters['operator_strength'] = self.SA_params['operator_strength']
        self.parameters['max_det_op'] = self.SA_params['max_det_op']
        self.parameters['consider_operator'] = self.SA_params['consider_operator']
        self.parameters['consider_daylight'] = self.SA_params['consider_daylight']
        self.parameters['consider_venting'] = self.SA_params['consider_venting']

        # Modify input distributions
        self.state['empirical_leaks'] = self.adjust_distribution(self.state['empirical_leaks'],
            self.SA_params['LSD_outliers'], self.SA_params['LSD_samples'])
        self.state['empirical_counts'] = self.adjust_distribution(self.state['empirical_counts'],
            self.SA_params['LCD_outliers'], self.SA_params['LCD_samples'])
        self.state['offsite_times'] = self.adjust_distribution(self.state['offsite_times'],
            self.SA_params['offsite_times_outliers'], self.SA_params['offsite_times_samples'])
        if self.parameters['consider_venting']:
            self.state['empirical_sites'] = self.adjust_distribution(self.state['empirical_sites'],
                self.SA_params['site_rate_outliers'], self.SA_params['site_rate_samples'])

        # Set OGI parameters                      
        if self.parameters['sensitivity']['program'] == 'OGI':
            self.parameters['methods']['OGI']['max_workday'] = self.SA_params['max_workday']
            self.parameters['methods']['OGI']['n_crews'] = self.SA_params['OGI_n_crews']
            self.parameters['methods']['OGI']['min_temp'] = self.SA_params['OGI_min_temp']
            self.parameters['methods']['OGI']['max_wind'] = self.SA_params['OGI_max_wind']
            self.parameters['methods']['OGI']['max_precip'] = self.SA_params['OGI_max_precip']
            self.parameters['methods']['OGI']['min_interval'] = self.SA_params['OGI_min_interval']
            self.parameters['methods']['OGI']['reporting_delay'] = self.SA_params['OGI_reporting_delay']
            self.parameters['methods']['OGI']['MDL'][0] = self.SA_params['OGI_MDL']

            for site in self.state['sites']:
                site['OGI_time'] = self.SA_params['OGI_time']
                site['OGI_required_surveys'] = self.SA_params['OGI_required_surveys']

        # Set screening (truck) parameters                      
        if self.parameters['sensitivity']['program'] == 'truck':
            self.parameters['methods']['OGI_FU']['max_workday'] = self.SA_params['max_workday']
            self.parameters['methods']['OGI_FU']['n_crews'] = self.SA_params['OGI_n_crews']
            self.parameters['methods']['OGI_FU']['min_temp'] = self.SA_params['OGI_min_temp']
            self.parameters['methods']['OGI_FU']['max_wind'] = self.SA_params['OGI_max_wind']
            self.parameters['methods']['OGI_FU']['max_precip'] = self.SA_params['OGI_max_precip']
            self.parameters['methods']['OGI_FU']['min_interval'] = self.SA_params['OGI_min_interval']
            self.parameters['methods']['OGI_FU']['reporting_delay'] = self.SA_params['OGI_reporting_delay']
            self.parameters['methods']['OGI_FU']['MDL'][0] = self.SA_params['OGI_MDL']

            self.parameters['methods']['truck']['max_workday'] = self.SA_params['max_workday']
            self.parameters['methods']['truck']['min_temp'] = self.SA_params['truck_min_temp']
            self.parameters['methods']['truck']['max_wind'] = self.SA_params['truck_max_wind']
            self.parameters['methods']['truck']['max_precip'] = self.SA_params['truck_max_precip']
            self.parameters['methods']['truck']['min_interval'] = self.SA_params['truck_min_interval']
            self.parameters['methods']['truck']['reporting_delay'] = self.SA_params['truck_reporting_delay']
            self.parameters['methods']['truck']['MDL'] = self.SA_params['truck_MDL']
            self.parameters['methods']['truck']['follow_up_thresh'] = self.SA_params['truck_follow_up_thresh']
            self.parameters['methods']['truck']['follow_up_ratio'] = self.SA_params['truck_follow_up_ratio']

            for site in self.state['sites']:
                site['OGI_time'] = self.SA_params['OGI_time']
                site['OGI_FU_time'] = self.SA_params['OGI_time']
                site['OGI_required_surveys'] = self.SA_params['OGI_required_surveys']
                site['truck_time'] = self.SA_params['truck_time']
                site['truck_required_surveys'] = self.SA_params['truck_required_surveys']

        return

    # ------------------------------------------------------------------------------

    def write_data(self):

        # Build generic output dataframe
        df_dict = {
            # SA inputs
            'program': self.parameters['sensitivity']['program'],
            'simulation': self.parameters['simulation'],
            'run_time': time.time() - self.parameters['start_time'],
            'timesteps': self.parameters['timesteps'],
            'spin_up': self.parameters['spin_up'],
            'start_year': self.parameters['start_year'],
            'LSD_outliers': self.SA_params['LSD_outliers'],
            'LSD_samples': self.SA_params['LSD_samples'],
            'LCD_outliers': self.SA_params['LCD_outliers'],
            'LCD_samples': self.SA_params['LCD_samples'],
            'site_rate_outliers': self.SA_params['site_rate_outliers'],
            'site_rate_samples': self.SA_params['site_rate_samples'],
            'offsite_times_outliers': self.SA_params['offsite_times_outliers'],
            'offsite_times_samples': self.SA_params['offsite_times_samples'],
            'LPR': self.SA_params['LPR'],
            'repair_delay': self.SA_params['repair_delay'],
            'operator_strength': self.SA_params['operator_strength'],
            'max_det_op': self.SA_params['max_det_op'],

            # SA outputs
            'dail_site_em': np.mean(np.array(self.timeseries['daily_emissions_kg'][self.parameters['spin_up']:]) / len(
                self.state['sites'])),
            'std_dail_site_em': np.std(
                np.array(self.timeseries['daily_emissions_kg'][self.parameters['spin_up']:]) / len(
                    self.state['sites'])),
            'cum_repaired_leaks': self.timeseries['cum_repaired_leaks'][-1:][0],
            'med_active_leaks': np.median(np.array(self.timeseries['active_leaks'][self.parameters['spin_up']:])),
            'med_days_active': np.median(pd.DataFrame(self.state['leaks'])['days_active']),
            'med_leak_rate': np.median(pd.DataFrame(self.state['leaks'])['rate']),
            'med_vent_rate': np.median(self.state['empirical_vents']),
            'cum_init_leaks': np.sum(pd.DataFrame(self.state['sites'])['initial_leaks']),
        }

        # If this is an operator program, you're done - export the data
        if self.parameters['sensitivity']['program'] == 'operator':
            self.export_SA(df_dict, self.output_directory, 'sensitivity_operator.csv')

            # Otherwise, this is an LDAR program - add generic inputs
        generic_dict = {
            'max_workday': self.SA_params['max_workday'],
            'consider_operator': self.SA_params['consider_operator'],
            'consider_daylight': self.SA_params['consider_daylight'],
            'consider_venting': self.SA_params['consider_venting']
        }
        df_dict.update(generic_dict)

        # If this is an OGI program, add relevant variables to dictionary before exporting
        if self.parameters['sensitivity']['program'] == 'OGI':
            OGI_dict = {
                # New OGI inputs
                'OGI_n_crews': self.SA_params['OGI_n_crews'],
                'OGI_min_temp': self.SA_params['OGI_min_temp'],
                'OGI_max_wind': self.SA_params['OGI_max_wind'],
                'OGI_max_precip': self.SA_params['OGI_max_precip'],
                'OGI_min_interval': self.SA_params['OGI_min_interval'],
                'OGI_reporting_delay': self.SA_params['OGI_reporting_delay'],
                'OGI_MDL': self.SA_params['OGI_MDL'],
                'OGI_time': self.SA_params['OGI_time'],
                'OGI_required_surveys': self.SA_params['OGI_required_surveys'],

                # New OGI outputs
                'OGI_cum_program_cost': np.sum(np.array(self.timeseries['OGI_cost'][self.parameters['spin_up']:])),
                'OGI_prop_sites_avail': np.mean(
                    np.array(self.timeseries['OGI_prop_sites_avail'][self.parameters['spin_up']:])),
                'OGI_cum_missed_leaks': np.sum(pd.DataFrame(self.state['sites'])['OGI_missed_leaks']),
                'OGI_cum_surveys': np.sum(pd.DataFrame(self.state['sites'])['OGI_surveys_conducted'])
            }

            df_dict.update(OGI_dict)
            self.export_SA(df_dict, self.output_directory, 'sensitivity_OGI.csv')

        # If this is a screening (truck) program, add relevant variables to dictionary before exporting
        if self.parameters['sensitivity']['program'] == 'truck':
            truck_dict = {
                # New OGI_FU inputs
                'OGI_FU_n_crews': self.SA_params['OGI_n_crews'],
                'OGI_FU_min_temp': self.SA_params['OGI_min_temp'],
                'OGI_FU_max_wind': self.SA_params['OGI_max_wind'],
                'OGI_FU_max_precip': self.SA_params['OGI_max_precip'],
                'OGI_FU_min_interval': self.SA_params['OGI_min_interval'],
                'OGI_FU_reporting_delay': self.SA_params['OGI_reporting_delay'],
                'OGI_FU_MDL': self.SA_params['OGI_MDL'],
                'OGI_FU_time': self.SA_params['OGI_time'],

                # New truck inputs
                'truck_n_crews': self.SA_params['truck_n_crews'],
                'truck_min_temp': self.SA_params['truck_min_temp'],
                'truck_max_wind': self.SA_params['truck_max_wind'],
                'truck_max_precip': self.SA_params['truck_max_precip'],
                'truck_min_interval': self.SA_params['truck_min_interval'],
                'truck_reporting_delay': self.SA_params['truck_reporting_delay'],
                'truck_MDL': self.SA_params['truck_MDL'],
                'truck_follow_up_thresh': self.SA_params['truck_follow_up_thresh'],
                'truck_follow_up_ratio': self.SA_params['truck_follow_up_ratio'],
                'truck_time': self.SA_params['truck_time'],
                'truck_required_surveys': self.SA_params['truck_required_surveys'],

                # New OGI_FU outputs
                'OGI_FU_cum_program_cost': np.sum(
                    np.array(self.timeseries['OGI_FU_cost'][self.parameters['spin_up']:])),
                'OGI_FU_prop_sites_avail': np.mean(
                    np.array(self.timeseries['OGI_FU_prop_sites_avail'][self.parameters['spin_up']:])),
                'OGI_FU_cum_missed_leaks': np.sum(pd.DataFrame(self.state['sites'])['OGI_FU_missed_leaks']),
                'OGI_FU_cum_surveys': np.sum(pd.DataFrame(self.state['sites'])['OGI_FU_surveys_conducted']),

                # New truck outputs
                'truck_cum_program_cost': np.sum(np.array(self.timeseries['truck_cost'][self.parameters['spin_up']:])),
                'truck_prop_sites_avail': np.mean(
                    np.array(self.timeseries['truck_prop_sites_avail'][self.parameters['spin_up']:])),
                'truck_cum_missed_leaks': np.sum(pd.DataFrame(self.state['sites'])['truck_missed_leaks']),
                'truck_cum_surveys': np.sum(pd.DataFrame(self.state['sites'])['truck_surveys_conducted']),
                'truck_cum_eff_flags': np.sum(np.array(self.timeseries['truck_eff_flags'][self.parameters['spin_up']:]))
            }

            df_dict.update(truck_dict)
            self.export_SA(df_dict, self.output_directory, 'sensitivity_truck.csv')

        return (df_dict)

    def adjust_distribution(self, distribution, outliers, samples):
        if outliers < 0:
            for i in range(abs(outliers)):
                distribution = np.delete(distribution, np.where(distribution == max(distribution)))
        if outliers > 0:
            for i in range(outliers):
                new_value = max(distribution) * 2
                distribution = np.append(distribution, new_value)
        while samples < 10:
            samples = int(np.random.normal(len(distribution), len(distribution) / 4))
        distribution = np.random.choice(distribution, samples)
        return distribution

    def export_SA(self, dictionary, output_directory, name):
        if 'write_results_postsim' in self.parameters['sensitivity']:
            export_data = self.parameters['sensitivity']['write_results_postsim']
        else:
            export_data = True

        if export_data:
            df_new = pd.DataFrame([dictionary])
            output_file = os.path.join(output_directory, name)
            if not os.path.exists(output_file):
                df_new.to_csv(output_file, index=False)
            elif os.path.exists(output_file):
                for attempts in range(int(1e10)):
                    try:
                        df_old = pd.read_csv(output_file)
                        break
                    except: continue

                df_old = df_old.append(df_new)
                df_old.to_csv(output_file, index=False)
        return